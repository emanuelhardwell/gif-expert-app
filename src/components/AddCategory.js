import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
    console.log("handleInputChange ...");
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    // Se hace de esta forma ya que unicamente estamos pasando la FUNCION setCategories
    // y usamos un CALLBACK para tener la referencia de de las categorias en este caso
    // o bien podriamos pasar por las PROPS a CATEGORIES y quedaria asi ------ >
    //setCategories([...categories, "HunterXHunter"]);

    console.log("handleSubmit ...");
    console.log("---- " + inputValue + " ----");

    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    }
  };

  // RETURN
  return (
    <>
      <form onSubmit={handleSubmit}>
        <p>{inputValue}</p>
        <input
          className="form-control"
          type="text"
          name="buscar"
          placeholder="Buscar por categoria"
          value={inputValue}
          onChange={handleInputChange}
        />
      </form>
    </>
  );
};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
