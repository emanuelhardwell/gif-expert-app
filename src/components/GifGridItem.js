import React from "react";

export const GifGridItem = ({ id, title, url }) => {
  return (
    <>
      <div className="card">
        <img src={url} className="card-img-top" alt={title}></img>
        <div className="card-body">
          <p className="card-title"> {title} </p>
        </div>
      </div>
    </>
  );
};
