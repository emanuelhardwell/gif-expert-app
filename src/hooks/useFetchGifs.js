//
//

import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifs(category).then((gif) => {
      setTimeout(() => {
        console.log(gif);
        setState({
          data: gif,
          loading: false,
        });
      }, 3000);
    });
  }, [category]);

  return state;
};
