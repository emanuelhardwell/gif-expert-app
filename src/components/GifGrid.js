import React, { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  const [images, setImages] = useState([]);

  useEffect(() => {
    getGifs();
  }, []);

  const getGifs = async () => {
    const url =
      "https://api.giphy.com/v1/gifs/search?q=goku&limit=8&api_key=EMkLUTqDMhIksWGffC42PtHNNshPjgwj";
    const resp = await fetch(url);
    const { data } = await resp.json();
    // console.log(data);

    const gifs = data.map((image) => {
      return {
        id: image.id,
        title: image.title,
        url: image.images?.downsized?.url,
      };
    });

    console.log(gifs);
    setImages(gifs);
  };

  // RETURN
  return (
    <>
      <h3 className="text-center"> {category} </h3>
      <div className="container">
        <div className="row bg-dark">
          <div className="col-md-4 mb-3">
            {images.map((img) => (
              <GifGridItem key={img.id} {...img} />
            ))}
          </div>
        </div>
      </div>
    </>
  );
};
