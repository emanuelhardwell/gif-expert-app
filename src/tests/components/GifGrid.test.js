import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";

jest.mock("../../hooks/useFetchGifs");

//
describe("Pruebas en el componente GifGrid.js", () => {
  const category = "One Punch";

  test("should mostrar el snapshot correctamente", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should mostrar el snapshot correctamente, ya con una CATEGORY", () => {
    const gifs = [
      { id: "abc", title: "paco el chato", url: "http://example.com/paco.jpg" },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should mostrar el snapshot correctamente, ya con una CATEGORY... forma MAS ROBUSTA", () => {
    const gifs = [
      { id: "abc", title: "paco el chato", url: "http://example.com/paco.jpg" },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
