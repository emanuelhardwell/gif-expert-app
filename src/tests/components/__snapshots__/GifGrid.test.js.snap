// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente GifGrid.js should mostrar el snapshot correctamente 1`] = `
<Fragment>
  <h3
    className="text-center"
  >
     
    One Punch
     
  </h3>
  <p>
    Cargando data
  </p>
  <div
    className="container"
  >
    <div
      className="row bg-dark"
    >
      <div
        className="col-md-4 mb-3"
      />
    </div>
  </div>
</Fragment>
`;

exports[`Pruebas en el componente GifGrid.js should mostrar el snapshot correctamente, ya con una CATEGORY 1`] = `
<Fragment>
  <h3
    className="text-center"
  >
     
    One Punch
     
  </h3>
  <div
    className="container"
  >
    <div
      className="row bg-dark"
    >
      <div
        className="col-md-4 mb-3"
      >
        <GifGridItem
          id="abc"
          key="abc"
          title="paco el chato"
          url="http://example.com/paco.jpg"
        />
      </div>
    </div>
  </div>
</Fragment>
`;

exports[`Pruebas en el componente GifGrid.js should mostrar el snapshot correctamente, ya con una CATEGORY... forma MAS ROBUSTA 1`] = `
<Fragment>
  <h3
    className="text-center"
  >
     
    One Punch
     
  </h3>
  <div
    className="container"
  >
    <div
      className="row bg-dark"
    >
      <div
        className="col-md-4 mb-3"
      >
        <GifGridItem
          id="abc"
          key="abc"
          title="paco el chato"
          url="http://example.com/paco.jpg"
        />
      </div>
    </div>
  </div>
</Fragment>
`;
